### PR TITLE
Use a random hash instead of generating an url-based filename to store

### LINKED ISSUE #818: `store-response-dir`: Could not write response at path ... : file name too long
### httpx version:

v1.2.4

### Current Behavior:

When url is too long to be stored because of filepath length limitations, httpx fails to write response.

### Expected Behavior:

Use a random hash instead of generating an url-based filename to store.
As a complement, to be able to correlate written file with url, implement proposal of issue https://github.com/projectdiscovery/httpx/issues/817

### Steps To Reproduce:

`
echo 'https://www.google.com/search?q=long+url+example&hl=en&rls=en&ei=vmBMY5qoK5LFlwSAmJbABQ&ved=0ahUKEwia_9f5wuX6AhWS4oUKHQCMBVgQ4dUDCA0&uact=5&oq=long+url+example&gs_lcp=Cgdnd3Mtd2l6EAMyBQgAEIAEMgUIABCGAzIFCAAQhgMyBQgAEIYDOgoIABBHENYEELADOgcIABCwAxBDOgQIABBDOgYIABAWEB5KBAhBGABKBAhGGABQmQJY5Qtg5AxoAXABeACAAaQBiAGfCJIBAzAuN5gBAKABAcgBCsABAQ&sclient=gws-wiz' | httpx -follow-redirects -probe -json -no-fallback-scheme -srd /tmp/httpx.tmp | jq -r
`
```json
[ERR] Could not write response at path '/tmp/httpx.tmp/www.google.com.443[slash]search?q=long+url+example&hl=en&rls=en&ei=vmBMY5qoK5LFlwSAmJbABQ&ved=0ahUKEwia_9f5wuX6AhWS4oUKHQCMBVgQ4dUDCA0&uact=5&oq=long+url+example&gs_lcp=Cgdnd3Mtd2l6EAMyBQgAEIAEMgUIABCGAzIFCAAQhgMyBQgAEIYDOgoIABBHENYEELADOgcIABCwAxBDOgQIABBD.txt', to disk: open /tmp/httpx.tmp/www.google.com.443[slash]search?q=long+url+example&hl=en&rls=en&ei=vmBMY5qoK5LFlwSAmJbABQ&ved=0ahUKEwia_9f5wuX6AhWS4oUKHQCMBVgQ4dUDCA0&uact=5&oq=long+url+example&gs_lcp=Cgdnd3Mtd2l6EAMyBQgAEIAEMgUIABCGAzIFCAAQhgMyBQgAEIYDOgoIABBHENYEELADOgcIABCwAxBDOgQIABBD.txt: file name too long
{
  "timestamp": "2022-10-16T19:54:38.441887151Z",
  "hashes": {
    "body-md5": "ce27e98c6d70c49a980670157c92b22f",
    "body-mmh3": "1950973855",
    "body-sha256": "29c6a56ec3379a814d71f7ac958ae877d702a473f1e8977b592b40831c6327a9",
    "body-simhash": "15596775981373928395",
    "header-md5": "2bf9c262c8dfae7f3ede1f83bf61c39f",
    "header-mmh3": "-1658869733",
    "header-sha256": "626273cf89bc68e021c2b52229d23883196abb915f5298064015927bd59bba6d",
    "header-simhash": "9814072627145975727"
  },
  "port": "443",
  "url": "https://www.google.com:443/search?q=long+url+example&hl=en&rls=en&ei=vmBMY5qoK5LFlwSAmJbABQ&ved=0ahUKEwia_9f5wuX6AhWS4oUKHQCMBVgQ4dUDCA0&uact=5&oq=long+url+example&gs_lcp=Cgdnd3Mtd2l6EAMyBQgAEIAEMgUIABCGAzIFCAAQhgMyBQgAEIYDOgoIABBHENYEELADOgcIABCwAxBDOgQIABBDOgYIABAWEB5KBAhBGABKBAhGGABQmQJY5Qtg5AxoAXABeACAAaQBiAGfCJIBAzAuN5gBAKABAcgBCsABAQ&sclient=gws-wiz",
  "input": "https://www.google.com/search?q=long+url+example&hl=en&rls=en&ei=vmBMY5qoK5LFlwSAmJbABQ&ved=0ahUKEwia_9f5wuX6AhWS4oUKHQCMBVgQ4dUDCA0&uact=5&oq=long+url+example&gs_lcp=Cgdnd3Mtd2l6EAMyBQgAEIAEMgUIABCGAzIFCAAQhgMyBQgAEIYDOgoIABBHENYEELADOgcIABCwAxBDOgQIABBDOgYIABAWEB5KBAhBGABKBAhGGABQmQJY5Qtg5AxoAXABeACAAaQBiAGfCJIBAzAuN5gBAKABAcgBCsABAQ&sclient=gws-wiz",
  "title": "long url example - Google Search",
  "scheme": "https",
  "webserver": "gws",
  "content-type": "application/xhtml+xml",
  "method": "GET",
  "host": "142.250.72.164",
  "path": "/search?q=long+url+example&hl=en&rls=en&ei=vmBMY5qoK5LFlwSAmJbABQ&ved=0ahUKEwia_9f5wuX6AhWS4oUKHQCMBVgQ4dUDCA0&uact=5&oq=long+url+example&gs_lcp=Cgdnd3Mtd2l6EAMyBQgAEIAEMgUIABCGAzIFCAAQhgMyBQgAEIYDOgoIABBHENYEELADOgcIABCwAxBDOgQIABBDOgYIABAWEB5KBAhBGABKBAhGGABQmQJY5Qtg5AxoAXABeACAAaQBiAGfCJIBAzAuN5gBAKABAcgBCsABAQ&sclient=gws-wiz",
  "response-time": "553.203383ms",
  "a": [
    "142.250.72.164",
    "2607:f8b0:4005:80d::2004"
  ],
  "words": 1354,
  "lines": 1,
  "status-code": 200,
  "content-length": 21528,
  "failed": false
}
```

### PR DESCRIPTION
**Pull request by LuitelSamikshya**
closes #818
Test:
```
echo 'https://www.google.com/search?q=long+url+example&hl=en&rls=en&ei=vmBMY5qoK5LFlwSAmJbABQ&ved=0ahUKEwia_9f5wuX6AhWS4oUKHQCMBVgQ4dUDCA0&uact=5&oq=long+url+example&gs_lcp=Cgdnd3Mtd2l6EAMyBQgAEIAEMgUIABCGAzIFCAAQhgMyBQgAEIYDOgoIABBHENYEELADOgcIABCwAxBDOgQIABBDOgYIABAWEB5KBAhBGABKBAhGGABQmQJY5Qtg5AxoAXABeACAAaQBiAGfCJIBAzAuN5gBAKABAcgBCsABAQ&sclient=gws-wiz' | ./httpx -follow-redirects -probe -json -no-fallback-scheme -srd temp/store | jq`
```

```json
{
  "timestamp": "2022-11-02T09:36:57.985867-05:00",
  "csp": {
    "domains": [
      "https:",
      "http:",
      "https://csp.withgoogle.com/csp/gws/cdt1"
    ]
  },
  "hash": {
    "body_md5": "703076697ad7b771c974834d230e60f5",
    "body_mmh3": "-644300001",
    "body_sha256": "26da22b9175d329a05d0fd071faecfa76f28221c45cb4ea752af8c873db8fac6",
    "body_simhash": "9809892618964940782",
    "header_md5": "bff8ac964e4cf59a80e15c28ad2cb5be",
    "header_mmh3": "-437207301",
    "header_sha256": "769a560324dae8fe1b056db5b118df68b7146a7904043c6d58209b85f87dfe63",
    "header_simhash": "9814140865586378174"
  },
  "port": "443",
  "url": "https://www.google.com:443/search?q=long+url+example&hl=en&rls=en&ei=vmBMY5qoK5LFlwSAmJbABQ&ved=0ahUKEwia_9f5wuX6AhWS4oUKHQCMBVgQ4dUDCA0&uact=5&oq=long+url+example&gs_lcp=Cgdnd3Mtd2l6EAMyBQgAEIAEMgUIABCGAzIFCAAQhgMyBQgAEIYDOgoIABBHENYEELADOgcIABCwAxBDOgQIABBDOgYIABAWEB5KBAhBGABKBAhGGABQmQJY5Qtg5AxoAXABeACAAaQBiAGfCJIBAzAuN5gBAKABAcgBCsABAQ&sclient=gws-wiz",
  "input": "https://www.google.com/search?q=long+url+example&hl=en&rls=en&ei=vmBMY5qoK5LFlwSAmJbABQ&ved=0ahUKEwia_9f5wuX6AhWS4oUKHQCMBVgQ4dUDCA0&uact=5&oq=long+url+example&gs_lcp=Cgdnd3Mtd2l6EAMyBQgAEIAEMgUIABCGAzIFCAAQhgMyBQgAEIYDOgoIABBHENYEELADOgcIABCwAxBDOgQIABBDOgYIABAWEB5KBAhBGABKBAhGGABQmQJY5Qtg5AxoAXABeACAAaQBiAGfCJIBAzAuN5gBAKABAcgBCsABAQ&sclient=gws-wiz",
  "title": "long url example - Google Search",
  "scheme": "https",
  "webserver": "gws",
  "content_type": "text/html",
  "method": "GET",
  "host": "142.250.114.104",
  "path": "/search?q=long+url+example&hl=en&rls=en&ei=vmBMY5qoK5LFlwSAmJbABQ&ved=0ahUKEwia_9f5wuX6AhWS4oUKHQCMBVgQ4dUDCA0&uact=5&oq=long+url+example&gs_lcp=Cgdnd3Mtd2l6EAMyBQgAEIAEMgUIABCGAzIFCAAQhgMyBQgAEIYDOgoIABBHENYEELADOgcIABCwAxBDOgQIABBDOgYIABAWEB5KBAhBGABKBAhGGABQmQJY5Qtg5AxoAXABeACAAaQBiAGfCJIBAzAuN5gBAKABAcgBCsABAQ&sclient=gws-wiz",
  "time": "1.08023725s",
  "a": [
    "142.250.114.104",
    "142.250.114.105",
    "142.250.114.106",
    "142.250.114.147",
    "142.250.114.99",
    "142.250.114.103",
    "2607:f8b0:4023:1002::68",
    "2607:f8b0:4023:1002::69",
    "2607:f8b0:4023:1002::63",
    "2607:f8b0:4023:1002::67"
  ],
  "words": 14894,
  "lines": 248,
  "status_code": 200,
  "content_length": 499650,
  "failed": false,
  "stored_response_path": "**temp/store/fa3692d347cee373c38cdea83722f264.txt**"
}
```

### CODE REVIEW

**Review by ehsandeep — CHANGES_REQUESTED** [review]
The expected format for the output file name is - `endpoint_query_hash.txt`

Example: `https://www.google.com/search_XXXXXX.txt`

Note: windows has 255 max char limit for filenames

File: runner/options.go [hunk 1]
```diff
@@ -536,7 +535,9 @@ func (options *Options) ValidateOptions() error {
 		gologger.Debug().Msgf("Store response directory specified, enabling \"sr\" flag automatically\n")
 		options.StoreResponse = true
 	}
-
+	if options.StoreResponseDir != "" {
+		os.RemoveAll(path.Join(options.StoreResponseDir, "index.txt"))
```
Comment (Mzack9999) [inline runner/options.go:539]: `path` is for **URLs only**. 
Files requires `filepath` to be cross platform

**Review by Mzack9999 — CHANGES_REQUESTED** [review]

File: runner/options.go (hunk 1 above)
Comment (LuitelSamikshya) [reply runner/options.go]: Done

File: runner/runner.go [hunk 2]
```diff
@@ -1514,28 +1514,39 @@ retry:
...
 		// store response
-		responsePath = filepath.Join(scanopts.StoreResponseDirectory, domainFile)
+		responsePath = filepath.Join(domainBaseDir, domainFile)
 		respRaw := resp.Raw
+		reqRaw := requestDump
 		if len(respRaw) > scanopts.MaxResponseBodySizeToSave {
 			respRaw = respRaw[:scanopts.MaxResponseBodySizeToSave]
 		}
-		writeErr := os.WriteFile(responsePath, []byte(respRaw), 0644)
+		data := append([]byte(fmt.Sprintf("[%s]", fullURL)), append([]byte("\n\n"), reqRaw...)...)
```
Comment (Mzack9999) [inline runner/runner.go:1527]: Square brackets were just a placeholder `[full URL]`. I think `fmt.Sprintf("[%s]", fullURL)` => `fullURL`

File: runner/runner.go [hunk 3]
```diff
@@ -1514,28 +1514,39 @@ retry:
...
+		data = append(data, append([]byte("\n"), respRaw...)...)
+		_ = fileutil.CreateFolder(domainBaseDir)
+		writeErr := os.WriteFile(responsePath, data, 0644)
 		if writeErr != nil {
 			gologger.Error().Msgf("Could not write response at path '%s', to disk: %s", responsePath, writeErr)
 		}
+		status := strings.ReplaceAll(strings.Split(resp.RawHeaders, "\n")[0], "HTTP/1.1 ", "")
+		status = strings.ReplaceAll(status, "HTTP/2 ", "")
+		indexData := fmt.Sprintf("%s %s (%s)\n", responsePath, fullURL, status)
```
Comment (Mzack9999) [inline runner/runner.go:1536]: Looks like `status` contains a trailing newline that needs to be removed before writing the data:
```
...
out\192.168.1.1\c5eb5a4cc76a5cdb16e79864b9ccd26c3553f0c396d0a21bafb7be71c1efcd8c.txt https://192.168.1.1:443 (200 OK
)
...
```

File: runner/runner.go [hunk 4]
```diff
@@ -1514,28 +1514,39 @@ retry:
...
+		indexPath := filepath.Join(scanopts.StoreResponseDirectory, "index.txt")
+		file, err := os.OpenFile(indexPath, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
```
Comment (Mzack9999) [inline runner/runner.go:1538]: opening/closing the same file continously doesn't seem very efficient. We can open it once at the beginning, then just write to it for example with a `bufio.writer`.

**Review by Mzack9999 — CHANGES_REQUESTED** [review]

File: runner/runner.go (hunk 2 above)
Comment (LuitelSamikshya) [reply runner/runner.go]: Done.

File: runner/runner.go (hunk 3 above)
Comment (LuitelSamikshya) [reply runner/runner.go:1545]: Done

File: runner/runner.go (hunk 4 above)
Comment (LuitelSamikshya) [reply runner/runner.go]: used strings.Builder to prevent continuous  opening and closing of same file

File: runner/options.go [hunk 5]
```diff
@@ -255,6 +253,7 @@ type Options struct {
 	OutputFilterCondition     string
 	OutputMatchCondition      string
 	OnResult                  OnResultCallback
+	indexFileData             *strings.Builder
```
Comment (Mzack9999) [inline runner/options.go:256]: With large lists of input targets, the memory occupation will grow too much. We can open the output index file and sync-write to it, just like we do in the case of the output file (`-o` option)

**Review by Mzack9999 — CHANGES_REQUESTED** [review]

**Review by Mzack9999 — APPROVED** [review]
```console
$ cat list.txt | go run . -srd test
...
$ cat test/index.txt
test/hackerone.com/b12ceb125f2e005739a0b6d87c900aefc0f97e1ba22609e10ed81ffd9c7b50a8.txt https://hackerone.com:443 (302 Found)
test/projectdiscovery.io/1fca6224ecf596635299a271b83d8a6f060c1f43ba3473895045a5a0cdeea177.txt https://projectdiscovery.io:443 (200 OK)
test/192.168.1.1/c5eb5a4cc76a5cdb16e79864b9ccd26c3553f0c396d0a21bafb7be71c1efcd8c.txt https://192.168.1.1:443 (200 OK)
```

**Review by ehsandeep — APPROVED** [review]

**Comment by ehsandeep** [comment]
```
[✘] Test "scanme.sh {{binary}} -silent -l test-data/request.txt" failed: [https://scanme.sh/] main is not equal to [https://scanme.sh/ http://scanme.sh]/ dev
```

@Mzack9999

**Pull request merged**